Langchain labs

krab7191 opened the pull request:
- Basic introduction based on the dedicated squad lab
- Summarization lab based on various and sundry docs across the interwebs

jbukuts reviewed (commented):
Cool lab. I wasn't too familiar with langchain so it's cool to see something like map-reduce implemented in this lab.  

I will note I ran into a Python issue around Chroma DB similar to what's outlined here:

https://github.com/langchain-ai/langchain/issues/7482

They noted a fix had been merged into a newer version so I bumped my `langchain` version from `0.0.236` to `0.0.312` and had no issues after that. I was using Python version 3.11.5.

I'm not sure if this warrants an update to the `requirements.txt` file though. 

Otherwise LGTM.

timroster commented:
I recently updated the langchain dependency for the internal bootcamp to `0.0.312` due to the security issues with previous versions. This tested successfully across all of those labs. I think it may be a good plan to test and bump this in `requirements.txt` .

ridgill commented on content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:
```diff
@@ -0,0 +1,345 @@
+    "    GenParams.TEMPERATURE: 0.2,\n",
+    "    GenParams.TOP_P: 1,\n",
+    "    GenParams.TOP_K: 25,\n",
+    "    # GenParams.RANDOM_SEED: 42,\n",
```
Minor: should we remove the comment?

ridgill commented on content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb:
```diff
@@ -0,0 +1,313 @@
+   "source": [
+    "## 1. Summarization with langchain\n",
+    "\n",
+    "Summarization of long documents is a common LLM use case. The issue that most often arises, however is that there is a token limit for the model. (Max context window length) With langchain this can be worked around by chunking and recursive summarization."
```
Also, should the period should come after the "(Max content window length)"?

tkmonson commented on content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:
```diff
@@ -0,0 +1,345 @@
+    "In the previous example, the user input is sent directly to the Watsonx LLM, without using Langchain. This is a basic use case, but real applications are rarely so simple. When using an LLM in an application, you will usually need to reuse the same prompt across multiple scenarios. We will now replicate the previous example, but use an LLM chain. This allows us to:\n",
+    "\n",
+    "- Accept user input and contruct a prompt\n",
+    "- Generate mutiple prompts from an collection of data points in a dataset "
```
"an collection" -> "a collection"

tkmonson commented on content/Watsonx/WatsonxAI/105/langchain-intro.ipynb:
```diff
@@ -0,0 +1,345 @@
+   "metadata": {},
+   "source": [
+    "## 4. Easy Loading of Documents Using Lang Chain\n",
+    "LangChain makes it easy to extract passages from documents so that you can answering questions based on your document's content."
```
"answering" -> "answer"

ridgill commented on content/Watsonx/WatsonxAI/105/langchain-summarization.ipynb:
```diff
@@ -0,0 +1,313 @@
+    "reduce_chain = LLMChain(llm=flan_model, prompt=reduce_prompt)\n",
+    "\n",
+    "# Takes a list of documents, combines them into a single string, and passes this to an LLMChain\n",
+    "print(\"Stuff docuements using reduce chain...\")\n",
```
"documents"

tkmonson reviewed (commented):
Everything worked pretty smoothly for me. Some of the outputs of the models were not very good, but not sure that we can do much about that. I got some false answers, and the summarization output from the map-reduce document chain example was not good (included the titles of the articles, which is not helpful).

ridgill commented:
Awesome lab!  The only hiccup I encountered was the very last step in the summarization notebook.  
The first time I ran it there seemed to be no output beyond this one line.  Running a second time gives a more reasonable output.  Including both screenshots.  Not sure what would account for the first output:

Output1:
<img width="1299" alt="Screenshot 2023-10-12 at 1 17 17 PM" src="https://github.com/ibm-build-lab/VAD-VAR-Workshop/assets/7364498/4ef381e1-fc28-411a-b110-0385fc7bd5cb">

Output2:
<img width="1308" alt="Screenshot 2023-10-12 at 1 18 51 PM" src="https://github.com/ibm-build-lab/VAD-VAR-Workshop/assets/7364498/95e76a67-3412-4082-8d32-c3297f2c1cc9">

krab7191 commented:
- Fixed typos
- Added print out for summarization intermediate steps. Note: the chunking and tokenization seems to be the weak link here. I'll continue trying options around that to get a more consistent / better output.

ridgill reviewed (approved):
LGTM!
